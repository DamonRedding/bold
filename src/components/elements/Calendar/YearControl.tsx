import * as React from 'react'

import { getUserLocale } from '../../../util/locale'
import { HFlow } from '../../layout/Flow/HFlow'
import { Button } from '../Button'

export interface YearControlProps {
    visibleDate: Date
    onChange(newDate: Date): void
    renderYear?(date: Date): React.ReactNode
}

const formatter = new Intl.DateTimeFormat(getUserLocale(), {
    year: 'numeric',
})

export class YearControl extends React.PureComponent<YearControlProps> {

    static defaultProps: Partial<YearControlProps> = {
        renderYear: (date) => formatter.format(date),
    }

    render() {
        const { visibleDate, renderYear } = this.props
        return (
            <HFlow alignItems='center' hSpacing={0.5}>
                <Button icon='angleLeft' size='small' skin='ghost' onClick={this.handlePrev} tabIndex={-1} />
                {renderYear(visibleDate)}
                <Button icon='angleRight' size='small' skin='ghost' onClick={this.handleNext} tabIndex={-1} />
            </HFlow>
        )
    }

    handleNext = () => {
        const { visibleDate } = this.props
        const next = new Date(visibleDate)
        next.setFullYear(visibleDate.getFullYear() + 1)
        return this.props.onChange(next)
    }

    handlePrev = () => {
        const { visibleDate } = this.props
        const prev = new Date(visibleDate)
        prev.setFullYear(prev.getFullYear() - 1)
        return this.props.onChange(prev)
    }
}
