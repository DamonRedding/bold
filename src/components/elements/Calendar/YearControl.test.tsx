import { mount, shallow } from 'enzyme'
import React from 'react'

import { withTheme } from '../../../test'

import { YearControl } from './YearControl'

it('should render correctly', () => {
    expect(shallow(
        <YearControl visibleDate={new Date('2018-10-26')} onChange={jest.fn()} />
    )).toMatchSnapshot()
})

it('should call onChange function with next year when clicked on next button', () => {
    const change = jest.fn()
    const wrapper = mount(withTheme(
        <YearControl visibleDate={new Date('2018-10-26')} onChange={change} />
    ))
    expect(change).not.toHaveBeenCalled()
    wrapper.find('button').last().simulate('click')
    expect(change).toHaveBeenCalledWith(new Date('2019-10-26'))
})

it('should call onChange function with previous year when clicked on prev button', () => {
    const change = jest.fn()
    const wrapper = mount(withTheme(
        <YearControl visibleDate={new Date('2018-10-26')} onChange={change} />
    ))
    expect(change).not.toHaveBeenCalled()
    wrapper.find('button').first().simulate('click')
    expect(change).toHaveBeenCalledWith(new Date('2017-10-26'))
})
