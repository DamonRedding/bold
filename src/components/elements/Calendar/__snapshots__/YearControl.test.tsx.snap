// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<HFlow
  alignItems="center"
  hSpacing={0.5}
>
  <Button
    icon="angleLeft"
    onClick={[Function]}
    size="small"
    skin="ghost"
    tabIndex={-1}
  />
  2018
  <Button
    icon="angleRight"
    onClick={[Function]}
    size="small"
    skin="ghost"
    tabIndex={-1}
  />
</HFlow>
`;
